Zend/zend_compile.c

```diff
@@ -5776,6 +5776,9 @@ void zend_compile_params(zend_ast *ast, zend_ast *return_type_ast) /* {{{ */
 		}
 
 		ZEND_TYPE_FULL_MASK(arg_info->type) |= _ZEND_ARG_INFO_FLAGS(is_ref, is_variadic);
+		if (opcode == ZEND_RECV) {
+			opline->op2.num = ZEND_TYPE_FULL_MASK(arg_info->type);
```

Comment:
Looking at it again, my original example would get a performance hit if there were type coercions, because strict_types of the **calling** file is what gets checked, so the opcode handler would still have to check for string/float/bool->int.

- It still makes sense for array/object/nullable and no type
- If the absence of type coercion is much more common than type coercion in common applications/frameworks, the current approach of optimizing for the common case would still make sense.

```php
// fib(35) runtime decreased from .6842 to .6639 (3.0%)
function fib(int $n) : int {
    if ($n <= 1) {
        return $n;
    }
    return fib($n - 1) + fib($n - 2);
}
```

Reply:
@TysonAndre This line in your version of the patch is identical.